### ISSUE TITLE
Feature Request: support DeletionPolicy

### ISSUE THREAD

**Issue by huribeir** [aws/serverless-application-model#450]
This is a feature request to cover the implementation of support to using DeletionPolicy statements on AWS SAM resources. The result of specifying a DeletionPolicy attribute on a SAM resource should be the resulting CloudFormation resources also having that attribute

**Comment by shwetaskatdare** [comment]
Thanks for the request. We'll look into it.

**Comment by jfpaccini** [comment]
As per https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resource-attributes.html, only 'DependsOn' resource attribute is supported accross the board and DeletionPolicy is only supported for LayerVersion.

It would be very useful to support it for DynamoDb tables (aka AWS::Serverless::SimpleTable), as we do not want to delete a database table for some sensitive environments.

thanks

**Comment by PatMyron** [comment]
[`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html) and [`Metadata`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-metadata.html) as well

---

Without full support for [`DeletionPolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html) and [`UpdateReplacePolicy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html), [`AWS::Serverless` resources](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resources-and-properties.html) cannot be easily protected

Without support for [`Metadata`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-metadata.html), [resource level `cfn-lint` ignores](https://github.com/aws-cloudformation/cfn-python-lint/#resource-based-metadata) can't work, and these deployment safety Cloudformation Linter rules are forced to be ignored for entire templates, endangering other resource types as well:

https://github.com/aws-cloudformation/cfn-python-lint/issues/1265

**Comment by mhart** [comment]
I'm not sure if this is the same issue, but it looks like `UpdateReplacePolicy` is actually removed from an S3 bucket in the transformed output:

Original template:
```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:

  Bucket:
    Type: AWS::S3::Bucket
    DeletionPolicy: Retain
    UpdateReplacePolicy: Retain

  Function:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.handler
      Runtime: nodejs12.x
      Events:
        ObjectCreated:
          Type: S3
          Properties:
            Bucket: !Ref Bucket
            Events: s3:ObjectCreated:*
```

Transformed template (just the bucket portion):
```yaml
AWSTemplateFormatVersion: '2010-09-09'
Resources:
  Bucket:
    DeletionPolicy: Retain
    DependsOn:
    - FunctionObjectCreatedPermission
    Properties:
      NotificationConfiguration:
        LambdaConfigurations:
        - Event: s3:ObjectCreated:*
          Function:
            Fn::GetAtt:
            - Function
            - Arn
    Type: AWS::S3::Bucket
```

Note the `DeletionPolicy` is present, but the `UpdateReplacePolicy` has been removed.

Let me know if I should file a separate issue for this.

**Comment by rocifier** [comment]
How can I add `DeletionPolicy: Retain` onto the actual function itself?

**Comment by mndeveci** [comment]
Closing this issue since it is been released with [SAM v1.36.0](https://github.com/aws/serverless-application-model/releases/tag/v1.36.0) and [SAM CLI v1.24.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.24.0).

You can read more about how resource level attributes are handled in SAM via our updated documentation; https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resource-attributes.html

Thanks!
